File: ChessMaker/Scripts/Game/Move.js
function Move(player, piece, startPos, moveNum) {
    this.player = player;
    this.piece = piece;
    this.startPos = startPos;
    this.moveNumber = moveNum;
    this.notation = '';
    this.prevPieceMoveTurn = null;

    this.steps = [];
    this.references = {};
}

Move.prototype.addStep = function(step) {
    this.steps.push(step);
};

Move.prototype.clone = function() {
    var move = new Move(this.player, this.piece, this.startPos, this.moveNumber);
		
    for ( var i=0; i<this.steps.length; i++ )
        move.addStep(this.steps[i]);

    for (var ref in this.references) {
        move.addPieceReference(ref, this.references[ref]);
    }

    return move;
};

Move.prototype.perform = function (game, updateDisplay) {
    for (var i = 0; i < this.steps.length; i++) {
        if (!this.steps[i].perform(game, updateDisplay)) // move failed, roll-back
        {
            for (var j = i - 1; j >= 0; j--)
                if (!this.steps[j].reverse(game, updateDisplay))
                    throw "Move failed on step " + i + "/" + this.steps.length + ", and rolling move back then failed on step " + j + ". Unable to rectify board state.";

            return false;
        }
    }

    this.prevPieceMoveTurn = this.piece.lastMoveTurn;
    this.piece.lastMoveTurn = game.moveNumber;
    this.piece.moveNumber++;
    game.moveNumber++;
    return true;
};

Move.prototype.reverse = function (game, updateDisplay) {
    for (var i = this.steps.length - 1; i >= 0; i--)
        if (!this.steps[i].reverse(game, updateDisplay)) // move failed, roll-back
        {
            for (var j = i + 1; j < this.steps.length; j++)
                if (!this.steps[j].perform(game, updateDisplay))
                    throw "Reversing move failed on step " + i + "/" + this.steps.length + " (counting backwards), and rolling move back then failed on step " + j + ". Unable to rectify board state.";

            return false;
        }

    this.piece.lastMoveTurn = this.prevPieceMoveTurn;
    this.piece.moveNumber--;
    game.moveNumber--;
    return true;
};

Move.prototype.addPieceReference = function(piece, ref) {
    this.references[ref] = piece;
};

Move.prototype.getPieceByRef = function (ref) {
    if (ref == 'self')
        return this.piece;
    var piece = this.references[ref];
    if (piece === undefined)
        return null;
    return piece;
};

Move.prototype.getEndPos = function () {
    for (var i = this.steps.length - 1; i >= 0; i--) {
        var s = this.steps[i];
        if (s.piece == this.piece && s.toState == Piece.State.OnBoard)
            return s.toPos;
    }// if this piece doesn't move, consider other pieces that it might move or capture
    for (var i = this.steps.length - 1; i >= 0; i--) {
        var s = this.steps[i];
        if (s.toState == Piece.State.OnBoard)
            return s.toPos;
        else if (s.fromState == Piece.State.OnBoard)
            return s.fromPos;
    }
    return this.startPos;
};

Move.prototype.getAllPositions = function () {
    var allPos = [];
    for (var i = 0; i < this.steps.length; i++)
        if (this.steps[i].piece == this.piece && this.steps[i].toState == Piece.State.OnBoard)
            allPoints.push(this.steps[i].toPos);
    return allPos;
};

Move.prototype.isCapture = function () {
    for (var i = 0; i < this.steps.length; i++) {
        var s = this.steps[i];
        if (s.piece != this.piece && s.toState != Piece.State.OnBoard)
            return true;
    }
    return false;
};

Move.prototype.getPromotionType = function () {
    for (var i = this.steps.length - 1; i >= 0; i--) {
        var s = this.steps[i];
        if (s.piece == this.piece && s.toType != s.fromType)
            return s.fromType;
    }
    return null;
};

Move.maxNotationDetail = 3;
Move.prototype.determineNotation = function (detailLevel) {
    // include an X if capturing a piece
    var capture = this.isCapture() ? 'x' : '';

    var desc;
    switch (detailLevel) {
        case 1:
            desc = this.piece.pieceType.notation + capture + this.getEndPos().name;
            break;
        case 2:
            desc = this.piece.pieceType.notation + this.startPos.name + capture + this.getEndPos().name;
            break;
        default:
            desc = this.piece.pieceType.notation + '(' + this.piece.elementID + ')' + this.startPos.name + capture + this.getEndPos().name;
            break;
    }

    // add promotion letter onto the end
    var promotion = this.getPromotionType();
    if (promotion != null)
        desc += promotion.notation;

    // add a plus to the end if its check. add a hash if its checkmate.

    // how to account for castling? some special moves simply need to specify their own special notation, i guess...

    var displayNumber = Math.ceil(this.moveNumber / game.players.length);
    if (this.moveNumber % game.players.length == 1) // first player's move
        displayNumber += '.  ';
    else
        displayNumber += '...';

    this.notation = displayNumber + desc;
    return this.notation;
};

File: ChessmakerCore/Move.cpp
#include "Move.h"
#include "MoveStep.h"
#include "Game.h"
#include "GameState.h"
#include "TurnOrder.h"

Move::Move(Player *player, GameState *prevState, Piece *piece, Cell *startPos)
{
	this->player = player;
	this->prevState = prevState;
	this->piece = piece;
	this->startPos = startPos;
	hasCustomNotation = false;
	appendNotation[0] = 0;
}


Move::~Move()
{
	while (!steps.empty())
		delete steps.front(), steps.pop_front();
}


Move *Move::Clone()
{
	Move *move = new Move(player, prevState, piece, startPos);
	for (auto it = steps.begin(); it != steps.end(); it++)
	{
		MoveStep *step = *it;
		move->AddStep(new MoveStep(*step), false);
	}

	move->references = references;
	move->description = description;
	if (hasCustomNotation)
		move->notation = notation;
	move->hasCustomNotation = hasCustomNotation;
	strcpy(move->appendNotation, appendNotation);

	return move;
}


GameState *Move::Perform(bool updateDisplay)
{
	for (auto it = steps.begin(); it != steps.end(); it++)
	{
		MoveStep *step = *it;

		if (!step->Perform(updateDisplay)) // move failed, roll back
		{
			if (it == steps.begin())
				return 0;

			bool reversing = true;
			do
			{
				it--;
				reversing = it != steps.begin();

				step = *it;
				if (!step->Reverse(updateDisplay))
				{
					ReportError("Encountered an error rolling back a failed move\n");
					return 0;
				}
			} while (reversing);
			
			return 0;
		}
	}

	prevPieceMoveTurn = piece->lastMoveTurn;
	piece->lastMoveTurn = prevState->turnNumber;
	piece->moveNumber++;
	prevState->subsequentMove = this;

	return new GameState(prevState->game, prevState->game->GetTurnOrder()->GetNextPlayer(), prevState);
}


bool Move::Reverse(bool updateDisplay)
{
	piece->lastMoveTurn = prevPieceMoveTurn;
	piece->moveNumber--;
	prevState->game->GetTurnOrder()->StepBackward();
	if (prevState->subsequentMove == this)
		prevState->subsequentMove = 0;

	for (auto it = steps.rbegin(); it != steps.rend(); it++)
	{
		MoveStep *step = *it;

		if (!step->Reverse(updateDisplay)) // move failed, roll back
		{
			if (it == steps.rbegin())
				return false;

			bool reversing = true;
			do
			{
				it--;
				reversing = it != steps.rbegin();

				step = *it;
				if (!step->Perform(updateDisplay))
				{
					ReportError("Encountered an error rolling back a failed move reversal\n");
					return false;
				}
			} while (reversing);

			return false;
		}
	}

	return true;
}


const char *Move::DetermineNotation(int detailLevel)
{
	// if I already have a notation set, just return that?
	if (hasCustomNotation)
		return notation.c_str();

    switch (detailLevel)
	{
        case 1:
			notation = piece->pieceType->GetNotation();
			
			if (IsDrop())
				notation += "*";
			
			if (IsCapture())
				notation += "x";
			
			notation += GetEndPos()->GetName();
            break;
        case 2:
			notation = piece->pieceType->notation;
			
			if (IsDrop())
				notation += "*";
			else
				notation += startPos->GetName();
			
			if (IsCapture())
				notation += "x";
			
			notation += GetEndPos()->GetName();
            break;
        default:
			notation = piece->pieceType->GetNotation();
			notation += "(";
			notation += std::to_string(piece->GetID());
			notation += ")";
			
			if (IsDrop())
				notation += "*";
			else
				notation += startPos->GetName(); 
			
			if (IsCapture())
				notation += "x";
			
			notation += GetEndPos()->GetName();
            break;
    }
	
	notation += appendNotation;
	return notation.c_str();
}


void Move::SetAppendNotation(const char *append)
{
	strcpy(appendNotation, append);
}


void Move::AddPieceReference(Piece *piece, const char *ref)
{
	auto it = references.find(ref);
	if (it != references.end())
		it->second = piece;
	else
		references.insert(std::pair<const char*, Piece*>(ref, piece));
}


Piece *Move::GetPieceByReference(const char *ref, Piece *self)
{
	if (strcmp(ref, "self") == 0)
		return self; // this is needed only by conditions code, it looks like. If they could be made to handle "self" separately, this wouldn't be needed.

	auto it = references.find(ref);
	if (it == references.end())
		return 0;
	return it->second;
}


Cell *Move::GetEndPos()
{
	for (auto it = steps.rbegin(); it != steps.rend(); it++)
	{
		MoveStep *step = *it;
		if (step->GetPiece() == piece && step->toState == Piece::OnBoard)
			return step->toPos;
	}

	// if this piece doesn't move, consider other pieces that it might move or capture
	for (auto it = steps.rbegin(); it != steps.rend(); it++)
	{
		MoveStep *step = *it;
		if (step->toState == Piece::OnBoard)
			return step->toPos;
		else if (step->fromState == Piece::OnBoard)
			return step->fromPos;
	}

    return startPos;
}


std::list<Piece*> *Move::GetAllCaptures()
{
	std::list<Piece*> *allCaptures = new std::list<Piece*>();

	for (auto it = steps.begin(); it != steps.end(); it++)
	{
		MoveStep *step = *it;
		if (step->toState != Piece::OnBoard)
			allCaptures->push_back(piece);
	}
	return allCaptures;
}


bool Move::IsCapture()
{
	for (auto it = steps.begin(); it != steps.end(); it++)
	{
		MoveStep *step = *it;
		if (step->GetPiece() != piece && step->toState != Piece::OnBoard)
			return true;
	}
    return false;
}


bool Move::IsDrop()
{
	MoveStep *step = *steps.begin();
	return step->GetFromState() == Piece::Held;
}


bool Move::WouldCapture(Cell* target)
{
	for (auto it = steps.begin(); it != steps.end(); it++)
	{
		MoveStep *step = *it;
		if (step->GetPiece() != piece && step->toState != Piece::OnBoard && step->fromState == Piece::OnBoard && step->fromPos == target)
			return true;
	}
	return false;
}


PieceType *Move::GetPromotionType()
{
	for (auto it = steps.rbegin(); it != steps.rend(); it++)
	{
		MoveStep *step = *it;
		if (step->GetPiece() == piece && step->toType != step->fromType)
			return step->toType;
	}
	return 0;
}